File: other-programming-languages/Perl/2024/2024-005/2024-005.pl
use strict;
use warnings;
use List::Util qw(sum);

# Reading the file content
my $file_path = 'input.txt';
open(my $fh, '<', $file_path) or die "Cannot open file: $!";
my $content = do { local $/; <$fh> };
close($fh);

# Splitting content into rules and updates
my ($rules_section, $updates_section) = split(/\n\n/, $content);

my @rules = map { [split(/\|/, $_)] } split(/\n/, $rules_section);
@rules = map { [int($_->[0]), int($_->[1])] } @rules;

my @updates = map { [split(/,/, $_)] } split(/\n/, $updates_section);
@updates = map { [map { int($_) } @$_] } @updates;

# Helper function to check if an update follows the rules
sub is_update_ordered {
    my ($update, $rules) = @_;
    my %index_map = map { $update->[$_] => $_ } 0..$#$update;

    foreach my $rule (@$rules) {
        my ($x, $y) = @$rule;
        if (exists $index_map{$x} && exists $index_map{$y} && $index_map{$x} > $index_map{$y}) {
            return 0;
        }
    }
    return 1;
}

# Identify correctly ordered updates and their middle page numbers
my @correct_updates;
my @middle_pages;

foreach my $update (@updates) {
    if (is_update_ordered($update, \@rules)) {
        push @correct_updates, $update;
        push @middle_pages, $update->[int(@$update / 2)];
    }
}

# Calculate the sum of middle pages
my $sum_middle_pages = sum(@middle_pages);
print "Sum of middle pages for correctly ordered updates: $sum_middle_pages\n";

# Helper function to sort an update according to the rules
sub topological_sort {
    my ($update, $rules) = @_;
    my %graph;
    my %in_degree;
    my %nodes = map { $_ => 1 } @$update;

    foreach my $rule (@$rules) {
        my ($x, $y) = @$rule;
        if (exists $nodes{$x} && exists $nodes{$y}) {
            push @{$graph{$x}}, $y;
            $in_degree{$y}++;
            $in_degree{$x} //= 0;
        }
    }

    my @queue = grep { $in_degree{$_} == 0 } keys %nodes;
    my @sorted_update;

    while (@queue) {
        my $current = shift @queue;
        push @sorted_update, $current;

        foreach my $neighbor (@{$graph{$current} || []}) {
            $in_degree{$neighbor}--;
            if ($in_degree{$neighbor} == 0) {
                push @queue, $neighbor;
            }
        }
    }

    return \@sorted_update;
}

# Correct the incorrectly ordered updates and find their middle page numbers
my @incorrect_updates;
my @incorrect_middle_pages;

foreach my $update (@updates) {
    unless (is_update_ordered($update, \@rules)) {
        my $corrected_update = topological_sort($update, \@rules);
        push @incorrect_updates, $corrected_update;
        push @incorrect_middle_pages, $corrected_update->[int(@$corrected_update / 2)];
    }
}

# Calculate the sum of middle pages for corrected updates
my $sum_incorrect_middle_pages = sum(@incorrect_middle_pages);
print "Sum of middle pages for corrected updates: $sum_incorrect_middle_pages\n";

File: other-programming-languages/Typescript/2024/2024-005/2024-005.ts
import * as fs from 'fs';

// Reading the file content
const filePath = "\\\\vmware-host\\Shared Folders\\C\\advent-of-code-002\\input-files\\2024\\2024-005\\input.txt";
const content = fs.readFileSync(filePath, 'utf8');

// Splitting content into rules and updates
const [rulesSection, updatesSection] = content.trim().split("\n\n");

const rules: [number, number][] = rulesSection.split("\n").map(line => {
    const [x, y] = line.split("|").map(Number);
    return [x, y];
});

const updates: number[][] = updatesSection.split("\n").map(line =>
    line.split(",").map(Number)
);

// Helper function to check if an update follows the rules
function isUpdateOrdered(update: number[], rules: [number, number][]): boolean {
    const indexMap: Record<number, number> = {};
    update.forEach((page, index) => {
        indexMap[page] = index;
    });

    for (const [x, y] of rules) {
        if (x in indexMap && y in indexMap && indexMap[x] > indexMap[y]) {
            return false;
        }
    }
    return true;
}

// Identify correctly ordered updates and their middle page numbers
const correctUpdates: number[][] = [];
const middlePages: number[] = [];

for (const update of updates) {
    if (isUpdateOrdered(update, rules)) {
        correctUpdates.push(update);
        middlePages.push(update[Math.floor(update.length / 2)]);
    }
}

// Calculate the sum of middle pages
const sumMiddlePages = middlePages.reduce((acc, val) => acc + val, 0);
console.log("Sum of middle pages for correctly ordered updates:", sumMiddlePages);

// Helper function to sort an update according to the rules
function topologicalSort(update: number[], rules: [number, number][]): number[] {
    const graph: Record<number, number[]> = {};
    const inDegree: Record<number, number> = {};
    const nodes = new Set(update);

    rules.forEach(([x, y]) => {
        if (nodes.has(x) && nodes.has(y)) {
            if (!graph[x]) graph[x] = [];
            graph[x].push(y);
            inDegree[y] = (inDegree[y] || 0) + 1;
            if (!(x in inDegree)) inDegree[x] = 0;
        }
    });

    const queue: number[] = Array.from(nodes).filter(node => (inDegree[node] || 0) === 0);
    const sortedUpdate: number[] = [];

    while (queue.length > 0) {
        const current = queue.shift()!;
        sortedUpdate.push(current);

        (graph[current] || []).forEach(neighbor => {
            inDegree[neighbor]--;
            if (inDegree[neighbor] === 0) queue.push(neighbor);
        });
    }

    return sortedUpdate;
}

// Correct the incorrectly ordered updates and find their middle page numbers
const incorrectUpdates: number[][] = [];
const incorrectMiddlePages: number[] = [];

for (const update of updates) {
    if (!isUpdateOrdered(update, rules)) {
        const correctedUpdate = topologicalSort(update, rules);
        incorrectUpdates.push(correctedUpdate);
        incorrectMiddlePages.push(correctedUpdate[Math.floor(correctedUpdate.length / 2)]);
    }
}

// Calculate the sum of middle pages for corrected updates
const sumIncorrectMiddlePages = incorrectMiddlePages.reduce((acc, val) => acc + val, 0);
console.log("Sum of middle pages for corrected updates:", sumIncorrectMiddlePages);
